src/sys/freebsd.rs

```diff
@@ -1 +1,268 @@
+    pub fn extattr_set_fd(fd: c_int, attrname: c_int, attrname: *const c_char, data: *const c_void, nbytes: size_t) -> ssize_t;
+
+    pub fn extattr_list_link(path: *const c_char, attrnamespace: c_int, data: *mut c_void, nbytes: size_t) -> ssize_t;
+    pub fn extattr_get_link(path: *const c_char, attrnamespace: c_int, attrname: *const c_char, data: *mut c_void, nbytes: size_t) -> ssize_t;
+    pub fn extattr_delete_link(path: *const c_char, attrname: c_int, attrname: *const c_char) -> c_int;
+    pub fn extattr_set_link(path: *const c_char, attrname: c_int, attrname: *const c_char, data: *const c_void, nbytes: size_t) -> ssize_t;
+}
+
+pub struct XAttrs {
+    user_attrs: Box<[u8]>,
+    system_attrs: Box<[u8]>,
+    offset: usize,
+}
+
+impl Clone for XAttrs {
+    fn clone(&self) -> Self {
+        XAttrs {
+            user_attrs: Vec::from(&*self.user_attrs).into_boxed_slice(),
+            system_attrs: Vec::from(&*self.system_attrs).into_boxed_slice(),
+            offset: self.offset
+        }
+    }
+
+    fn clone_from(&mut self, other: &XAttrs) {
+        self.offset = other.offset;
+
+        let mut data = mem::replace(&mut self.user_attrs, Box::new([])).into_vec();
+        data.extend(other.user_attrs.iter().cloned());
+        self.user_attrs = data.into_boxed_slice();
+
+        data = mem::replace(&mut self.system_attrs, Box::new([])).into_vec();
+        data.extend(other.system_attrs.iter().cloned());
+        self.system_attrs = data.into_boxed_slice();
+    }
+}
+
+impl Iterator for XAttrs {
+    type Item = OsString;
+    fn next(&mut self) -> Option<OsString> {
+        if self.user_attrs.is_empty() && self.system_attrs.is_empty() {
+            return None;
+        }
+
+        if self.offset == self.user_attrs.len() + self.system_attrs.len() {
+            return None;
+        }
+
+        let data = if self.offset < self.system_attrs.len() {
+            &self.system_attrs[self.offset..]
+        } else {
+            &self.user_attrs[self.offset - self.system_attrs.len() ..]
+        };
+
+        let siz = data[0] as usize;
+
+        self.offset += siz + 1;
+        if self.offset < self.system_attrs.len() {
+            Some(prefix_namespace(OsStr::from_bytes(&data[1..siz + 1]).to_owned(), EXTATTR_NAMESPACE_SYSTEM))
+        } else {
+            Some(prefix_namespace(OsStr::from_bytes(&data[1..siz + 1]).to_owned(), EXTATTR_NAMESPACE_USER))
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        if self.user_attrs.len() + self.system_attrs.len() == self.offset {
+            (0, Some(0))
+        } else {
+            (1, None)
+        }
+    }
+}
+
+fn name_to_ns(name: &OsStr) -> io::Result<(c_int, CString)> {
+    let name_str = match name.to_str() {
+        Some(s) => s,
+        None => return Err(io::Error::new(io::ErrorKind::InvalidInput, "cannot convert to rust string"))
+    };
+
+    let dot = name_str.find('.').unwrap();
```

Comment:
We shouldn't panic on invalid attribute names (names that don't contain a namespace).